### PR TITLE
Add Queue mode support at server side adding a PresenceService

### PR DESCRIPTION
**Pull request by sbernard31**

### CODE REVIEW

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+        listeners.remove(listener);
+    }
+
+    @Override
+    public boolean isClientAwake(Registration registration) {
+        if (!clientStatusList.containsKey(registration.getEndpoint())) {
+            return false;
+        }
+
+        PresenceStatus presenceStatus = clientStatusList.get(registration.getEndpoint());
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: Variable allocation not needed.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    public void setAwake(Registration reg) {
+        if (reg.usesQueueMode()) {
+            if (!clientStatusList.containsKey(reg.getEndpoint())) {
+                createPresenceStatusObject(reg);
+            }
+
+            boolean stateChanged = false;
+            synchronized (this) {
+                stateChanged = clientStatusList.get(reg.getEndpoint()).setAwake();
+                startClientAwakeTimer(reg);
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: needed in the synchronized block?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+    /* The state of the client: Awake or Sleeping */
+    private Presence state;
+
+    ScheduledFuture<?> clientScheduledFuture;
+
+    private int clientAwakeTime;
+
+    public PresenceStatus() {
+        this.state = Presence.SLEEPING;
+        this.clientAwakeTime = 93000; /* ms, default CoAP value */
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceStatus.java]: Not easy to change this value with the current implementation. Could be in a parameter of PresenceServiceImpl, and the constructor with argument could be used.

File: leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/servlet/ClientServlet.java [hunk 4]
```diff
@@ -201,6 +203,10 @@ private void handleException(Exception e, HttpServletResponse resp) throws IOExc
             LOG.warn("Thread Interrupted", e);
             resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
             resp.getWriter().append("Thread Interrupted:").append(e.getMessage()).flush();
+        } else if (e instanceof ClientSleepingException) {
+            LOG.warn("Client is sleeping", e);
+            resp.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
```
Comment (msangoi) [inline leshan-server-demo/src/main/java/org/eclipse/leshan/server/demo/servlet/ClientServlet.java]: bad request?

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java [hunk 5]
```diff
@@ -305,6 +306,14 @@ public LeshanServerBuilder disableSecuredEndpoint() {
         return this;
     }
 
+    /**
+     * deactivate PresenceService which tracks presence of devices using  LWM2M Queue Mode.
+     */
+    public LeshanServerBuilder disableQueueModeSupport() {
```
Comment (msangoi) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java]: Should be called by default in the LeshanClusterServer implementation.

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/QueueModeLwM2mRequestSender.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+ * 
+ * The Eclipse Public License is available at
+ *    http://www.eclipse.org/legal/epl-v10.html
+ * and the Eclipse Distribution License is available at
+ *    http://www.eclipse.org/org/documents/edl-v10.html.
+ * 
+ * Contributors:
+ *     RISE SICS AB - initial API and implementation
+ *******************************************************************************/
+package org.eclipse.leshan.server.californium.impl;
```
Comment (msangoi) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/impl/QueueModeLwM2mRequestSender.java]: Why not in leshan-core?

File: leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java [hunk 7]
```diff
@@ -305,6 +306,14 @@ public LeshanServerBuilder disableSecuredEndpoint() {
         return this;
     }
 
+    /**
+     * deactivate PresenceService which tracks presence of devices using  LWM2M Queue Mode.
```
Comment (msangoi) [inline leshan-server-cf/src/main/java/org/eclipse/leshan/server/californium/LeshanServerBuilder.java]: We should also document that the sender behaviour is changed.

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+
+    /**
+     * Set the state of the client identified by registration as {@link Presence#AWAKE}
+     * 
+     * @param reg the client's registration object
+     */
+    public void setAwake(Registration reg) {
+        if (reg.usesQueueMode()) {
+            if (!clientStatusList.containsKey(reg.getEndpoint())) {
+                createPresenceStatusObject(reg);
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: possible race condition here?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+            }
+
+            boolean stateChanged = false;
+            synchronized (this) {
```
Comment (msangoi) [inline leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: synchronized on the PresenceStatus is not enough?

File: leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java (hunk 2 above)
Comment (sbernard31) [reply leshan-server-core/src/main/java/org/eclipse/leshan/server/queue/PresenceServiceImpl.java]: I would like to post pone polish about threadsafeness for the next version but I will do a first check now.

**Comment by sbernard31** [comment]
I taked into account most of the remarks.
If no objection I plan to integrate this tomorrow in master.

**Comment by sbernard31** [comment]
This PR is now integrated in master.
Queue mode is activated by default.
It is possible to return to previous behavior, using `LeshanServerBuilder.disableQueueModeSupport`

**Pull request merged**